src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -130,11 +136,162 @@ QString DBusService::getLogs() {
+    addresses.append(ip.toString());
+  }
+
   return WgQuickProcess::run(
       op, config.m_privateKey, config.m_deviceIpv4Address,
       config.m_deviceIpv6Address, config.m_serverIpv4Gateway,
       config.m_serverIpv6Gateway, config.m_serverPublicKey,
       config.m_serverIpv4AddrIn, config.m_serverIpv6AddrIn,
-      config.m_allowedIPAddressRanges.join(", "), config.m_serverPort,
-      config.m_ipv6Enabled);
+      addresses.join(", "), config.m_serverPort, config.m_ipv6Enabled);
+}
+
+static inline bool parseEndpoint(struct sockaddr* endpoint,
+                                 const QString& address, int port) {
+  QString portString = QString::number(port);
+
+  struct addrinfo hints;
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_DGRAM;
+  hints.ai_protocol = IPPROTO_UDP;
+
+  struct addrinfo* resolved = nullptr;
+  int retries = 15;
+
+  for (unsigned int timeout = 1000000;;
+       timeout = std::min((unsigned int)20000000, timeout * 6 / 5)) {
+    int rv = getaddrinfo(address.toLocal8Bit(), portString.toLocal8Bit(),
+                         &hints, &resolved);
+    if (!rv) {
+      break;
+    }
+
+    /* The set of return codes that are "permanent failures". All other
+     * possibilities are potentially transient.
+     *
+     * This is according to https://sourceware.org/glibc/wiki/NameResolver which
+     * states: "From the perspective of the application that calls getaddrinfo()
+     * it perhaps doesn't matter that much since EAI_FAIL, EAI_NONAME and
+     * EAI_NODATA are all permanent failure codes and the causes are all
+     * permanent failures in the sense that there is no point in retrying
+     * later."
+     *
+     * So this is what we do, except FreeBSD removed EAI_NODATA some time ago,
+     * so that's conditional.
+     */
+    if (rv == EAI_NONAME || rv == EAI_FAIL ||
+#ifdef EAI_NODATA
+        rv == EAI_NODATA ||
+#endif
+        (retries >= 0 && !retries--)) {
+      logger.log() << "Failed to resolve the address endpoint";
+      return false;
+    }
+
+    logger.log() << "Trying again in" << (timeout / 1000000.0) << "seconds";
+    usleep(timeout);
+  }
+
+  if ((resolved->ai_family == AF_INET &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in)) ||
+      (resolved->ai_family == AF_INET6 &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in6))) {
+    memcpy(endpoint, resolved->ai_addr, resolved->ai_addrlen);
+    freeaddrinfo(resolved);
+    return true;
+  }
+
+  logger.log() << "Invalid endpoint" << address;
+  freeaddrinfo(resolved);
+  return false;
+}
+
+bool DBusService::switchServer(const Config& config) {
+  logger.log() << "Switching server";
+
+  wg_device* device = static_cast<wg_device*>(calloc(1, sizeof(*device)));
+  if (!device) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  auto guard = qScopeGuard([&] { wg_free_device(device); });
+
+  strncpy(device->name, WG_INTERFACE, IFNAMSIZ - 1);
```

Comment:
c++ question - we can't assign `WG_INTERFACE` to `-> name` because (a) it'll fail (b) it might change (c) other?

Reply:
You mean: `device->name = WG_INTERFACE`

That doesn't work because `name` is an array of `char`: https://github.com/WireGuard/wireguard-tools/blob/master/contrib/embeddable-wg-library/wireguard.h#L72

In C++, you cannot assign values to an array in this way. You must use a method (or do it by your own). `strncpy` is a standard C function and it's code is something like this (pseudo code):

```
void strncpy(char* s1, const char *s2, size_t s) {
  for (int i = 0; s1 && s1[i] && s2 && s2[i] && i < s; ++i) {
    s1[i] = s2[i]; 
  }
}
```

Reply:
Thank you.